apps/frontend/src/app/Database/DataManagers/ArtifactData.ts

```diff
@@ -1,292 +1,311 @@
+      //swap to prevArt <-> prevChar && newArt <-> newChar(outside of this if)
+
+      if (prevArt)
+        super.setCached(prevArt.id, { ...prevArt, location: prevChar?.key ? charKeyToLocCharKey(prevChar.key) : "" })
+      if (newChar)
+        this.database.chars.setEquippedArtifact(charKeyToLocCharKey(newChar.key), slotKey, newArt.id)
+      if (prevChar)
+        this.database.chars.setEquippedArtifact(charKeyToLocCharKey(prevChar.key), slotKey, prevArt?.id ?? "")
+    } else
+      newArt.location && this.database.chars.triggerCharacter(newArt.location, "update")
+    return newArt
+  }
+  deCache(artifact: ICachedArtifact): IArtifact {
+    const { setKey, rarity, level, slotKey, mainStatKey, substats, location, exclude, lock } = artifact
+    return { setKey, rarity, level, slotKey, mainStatKey, substats: substats.map(substat => ({ key: substat.key, value: substat.value })), location, exclude, lock }
+  }
+
+  new(value: IArtifact): string {
+    const id = generateRandomArtID(new Set(this.keys), this.deletedArts)
+    this.set(id, value)
+    return id
+  }
+  remove(key: string) {
+    const art = this.get(key)
+    if (!art) return
+    art.location && this.database.chars.setEquippedArtifact(art.location, art.slotKey, "")
+    this.deletedArts.add(key)
+    super.remove(key)
+  }
+  setProbability(id: string, probability?: number) {
+    const art = this.get(id)
+    if (art) this.setCached(id, { ...art, probability })
+  }
+  clear(): void {
+    super.clear()
+    this.deletedArts = new Set<string>()
+  }
+  importGOOD(good: IGOOD & IGO, result: ImportResult) {
+    result.artifacts.beforeMerge = this.values.length
+
+    // Match artifacts for counter, metadata, and locations
+    const artifacts = good.artifacts
+
+    if (!Array.isArray(artifacts) || !artifacts.length) {
+      result.artifacts.notInImport = this.values.length
+      return
+    }
+
+    result.artifacts.import = artifacts.length
+    const idsToRemove = new Set(this.values.map(a => a.id))
+    const hasEquipment = artifacts.some(a => a.location)
+    artifacts.forEach((a, i) => {
+      const art = this.validate(a)
+      if (!art) return result.artifacts.invalid.push(a)
+
+      let importArt = art
+      let importKey: string | undefined = (a as ICachedArtifact).id
+
+      let { duplicated, upgraded } = result.ignoreDups ? { duplicated: [], upgraded: [] } : this.findDups(art)
+      // Don't reuse dups/upgrades
+      duplicated = duplicated.filter(a => idsToRemove.has(a.id))
+      upgraded = upgraded.filter(a => idsToRemove.has(a.id))
+      if (duplicated[0] || upgraded[0]) {
+        // Favor upgrades with the same location, else use 1st dupe
+        const match = (hasEquipment && art.location && upgraded[0]?.location === art.location) ? upgraded[0] : (duplicated[0] || upgraded[0])
+        idsToRemove.delete(match.id)
+        if (duplicated[0]) result.artifacts.unchanged.push(art)
+        else if (upgraded[0]) result.artifacts.upgraded.push(art)
```

Comment:
Based on if-condition on line 100, one of lines 104 or 105 is guaranteed to run, so we can simplify this